### DISCUSSION TITLE
Usage as a myst template?

### DISCUSSION THREAD

**Discussion by rowanc1** [kai-tub/latex-beamer-pure-minimalistic#75] [Ideas]
Hi @kai-tub, I am working on some of the @executablebooks projects and came across this repository after some of your recent discussions there. :)

I was wondering if you might be interested in helping to make this template accessible from using myst content? We are pulling together a lot of various templates in the https://github.com/myst-templates organization, and hopefully this would be the first Beamer template?! 🚀 You can check out some of the templates to get an idea of what might be necessary, basically a `template.yml` file and some placeholders to define the content.

This would allow you to use MyST markdown, including JupyterNotebooks to create Beamer presentations.

I would be more than happy to help out, and use this as a use case for building out the documentation and tutorials. We have a start at the [beamer docs](https://js.myst.tools/jtex/create-a-beamer-template) as well as put together a [video tutorial walkthrough](https://js.myst.tools/jtex/create-a-latex-template) yesterday on creating an article template.

Let me know if you are interested in being involved!

**Comment by welcome** [comment]
Thanks for opening your first issue here! 💖
Be sure to follow the issue template.
Then the issue will be worked on in no time. 👍

**Comment by kai-tub** [comment]
Hey, I am trying to work on following through on my own comments and moved it to 'issue' to the discussion thread :D 

> hopefully this would be the first Beamer template?! 

Thank you very much for considering me/my template as a potential "first template". :+1: 

> This would allow you to use MyST markdown, including JupyterNotebooks to create Beamer presentations.

This is something that I definitely already thought about previously ;)

> We have a start at the [beamer docs](https://js.myst.tools/jtex/create-a-beamer-template)

A quick comment: Especially for these things it would be really helpful to automatically generate the PDF slides directly from the documentation itself and embed it into the page. Maybe you can have a look at the GitHub workflow from this workflow to kinda see how I tried to auto-generate images (since they can be easily embedded in markdown, note this project was created _pre_ MyST days, so there is probably an easier solution)?

---

BUT I would also like to comment on my experience working with beamer and creating this theme for full transparency.

LaTeX is by far _the_ standard for writing nice looking _documents_ but I personally find that beamer presentations are not up to that standard.
More often than not, you would like to 'quickly' change the position of an image, align multiple images in a 'fancy' way, add footnotes next to images for reference, do slight modifications to the layout to better suit your need, etc.
Especially when working with _other_ non-beamer people there is often no such thing as _quickly_ changing something.

Then there is the issue that I would expect your audience to _stumble_ over the missing _web features_ that are inherited from the PDF format. Like no interactive graphs, no easy way to playback/embed videos etc.

For these reasons, I moved away from beamer and to [reveal.js](https://revealjs.com/) when I present something on my own and gave up and worked with google slides when working with others.

Creating the template itself was very time-intensive and not something that I would expect a user to be able to 'customize'.

My point is that even after I put _way too much_ time into this template (and I still love it!) I cannot recommend working with it when I am part of a group.
I wanted to give you some insight into why this repository isn't that active and why I as a _creator_ and Beamer user was ultimately pushed away to other solutions.
If I could wish for a MyST to beamer or reveal.js "converter" I would prefer the reveal.js solution because than I can get the "richness" of the HTML interactivity and would assume that some HTML/CSS knowledge is at least better transferable than Beamer/LaTeX specifics so that the customization part is easier :sweat_smile: 

That said...

---

I will definitely have a look over the weekend and tell you if I am planning on tackling this project and if I would need some help along the way :)

Maybe that would help to popularize beamer for some quick informal presentations that mostly consist of bullet points?

**Reply by rowanc1** [reply]
I completely agree on your assessment on both latex as the tool of choice for technical documents, and that beamer isn't really there for the same professionalism. The reason I stopped on your template and made this comment is that it is the first beamer presentation I actually liked design wise! It is really obvious that you put a ton of care into this.

I also want to get to revealjs presentations, and hopefully soon. Perhaps this is also a place where we can work together for mystjs in the future. :) The reason I started with latex is that we have been putting a ton of work into jtex and latex converters over this last month, and I was hoping for a quick-ish win in that space.

I am glad you are up for playing around with this! Now I am going to go to the other thread and apologize for the state of the docs that wasted a bit of your time ... !

**Reply by rowanc1** [reply]
> I would prefer the reveal.js solution because than I can get the "richness" of the HTML interactivity

A follow up question: do you have pointers to really nice reveal templates?

**Reply by kai-tub** [reply]
> The reason I stopped on your template and made this comment is that it is the first beamer presentation I actually liked design wise! It is really obvious that you put a ton of care into this.

Thanks for the kind words :smile: 
Yes, it was my _full-hearted_ attempt at making Beamer slides _work_ for me.
But yeah, after trying to use them for a couple of presentation I came to conclusion I wrote above...

> I also want to get to revealjs presentations, and hopefully soon.

Great! It sounds like the _best_ fit at creating neat data-heavy presentations :)

> Now I am going to go to the other thread and apologize for the state of the docs that wasted a bit of your time ... !

No worries.  The docs said that everything was in the early stages.

> do you have pointers to really nice reveal templates?

Sadly, no, not really...
I also thought about creating a template for revealjs but I had to learn HTML/CSS first. :sweat_smile: 
Which is what I've been doing for the last few months...

---

Btw, as someone who never touched HTML/CSS and started to learn it without any prior knowledge, I can't believe how hard it is!
How can it be that semantic HTML seems to be like some mythical creature?
My experience is as follows:

1. Read Mozilla docs
2. Check HTML spec to better understand its (section vs. article vs. aside, for example) use
3. Look up some accessibility guides and learn that they don't match up with the spec or that other browsers ignore the spec to better match the 'common' mistakes from the real world (safari ignoring lists if `list-item` is None) etc.

Or how the document model is not a real thing that browsers support, even though it is in spec. (there should not be a need for h2..h6 anymore since everything _should_ be `<section><h1>` but nobody implemented it yet.

:face_exhaling: 

Sorry, felt the need to went here :D 

TL;DR: No, I don't know any 'nice' looking revealjs templates. I just used the defaults most of the time.

**Comment by kai-tub** [comment]
Ok, I've looked at the documentation and played a bit around.
I guess I have a draft `template.yml` file ready but am having quite a bit of trouble with the `myst` tool.
What is the recommended way of installing it, currently?
After jumping around for a while I found [the hint that mystjs is not a separate tool yet and that I need to install curvenote](https://js.myst.tools/guide/intro).
So I went on to install `curvenote`, which requires an API token. Said and done, and then tried to compile it with `myst build test.md` (`myst` from curvenote-cli) and it then required a `curvenote.yml` file. Ran `curvenote init` and re-running `myst build test.md` just seemed to install the webpage target and not the pdf.
Then I tried to render the template with `jtex render` (as mentioned in the docs) but that doesn't seem to be published yet, as the command was unknown.
So I started to skip through the video tutorial and got to the point where I could reproduce the same mistakes (now using `myst export` missing target folders broke the build) and could render the template and _try_ to start the compiling but it fails because the flat file structure _enforcement_ breaks my beamer code (my code assumes that logos are in a logo folder).

It would be nice if you could provide me with the current recommended way to install the bleeding edge of myst (or with the version that is in sync with the documentation), maybe some of my issues are/were already resolved?
But a first few comments on things I noticed and some issues that would be nice if you could help me out with:

- How to install most recent myst (repeating just for completeness)
- Do I really have to enforce a flat file structure? Is there a way around that?
  - I would like to keep the same folder structure as in the source repository and would like to assume that I can split up project files into individual directories.
  - Another reason is that I would like to manage the template with a submodule (yes, I know many devs aren't a fan of that but I would say that this would be the perfect use-case for an exception and I want to see how bad it gets :D ). Then it will be easier to update both projects at the same time if necessary
- Looking at the _beamer_ guide, I do not follow how I can provide options to the `frame` environment. I understand that the env will be created when it _sees_ the first heading, but when working with beamer you frequently want/have to enable/disable page-breaks, layout (switching to plain) etc. 
- How do I configure the LaTeX render engine? My template was designed for LuaLaTeX (and is also backwards compatible with pdfLaTeX) but not tested for XeLaTeX, which I don't really want to support if it doesn't immediately work out of the box
  - https://github.com/kai-tub/latex-beamer-pure-minimalistic/blob/master/multi_lang_examples/README.md here I provide some discussion about why I prefer LuaLaTeX over XeLaTeX
- The [beamer example on the webpage](https://js.myst.tools/jtex/create-a-beamer-template) shows a _bug_. Unordered lists should be translated to itemize and not enumerate.
  - On that note, how could I change the default item environment? For my template I created vfilleditems as a replacement to itemize. Or many users prefer (or have to) use `compactitem` instead of `itemize`. Where would these options be changed?
- Really liked working with `jtex` :+1: The checking tool is some fine piece of software :wink:
  - One _unexpected_ thing was that it was so 'chatty' when reporting issues, but didn't say anything when there weren't any mistakes. For such a _user-friendly_ tool, I would expect to see a nice message saying something like: "No errors found" or similar so that I know that it passed.

Those are my initial impressions and comments :)
If you have time to help me, I think we can get the beamer template done in no time :rocket:

**Reply by rowanc1** [reply]
Thanks @kai-tub for trying it out, and I am glad you eventually found the video and that was more helpful, and really sorry that it wasn't easier. There is certainly a mismatch between reality and the docs at the moment that we are rapidly trying to close.

The most up to date version of myst-cli, technically doesn't exist. And the `myst` command that was exported is actually just an aliased curvenote. Over this next week @fwkoch will be pulling this out into `myst-cli` and actually addressing the `myst build`, we were using the docs as a way to reason about the interface. :( There is an issue [here](https://github.com/executablebooks/mystjs/issues/61) that is probably quite relevant. Sorry for the confusion.

On a positive note, I am SUPER stoked by all the feedback, especially that knowing the constraints we put in about the template (flat directory) are a hard _no_ from you. We can change that, and mostly did it for simplicity in the copying to temp files, etc.

> How do I configure the LaTeX render engine?

There is some ability to do this through the `build: engine: ` syntax (undocumented, [example here](https://github.com/myst-templates/egu_copernicus/blob/main/template.yml#L16)), but you probably may want more control. I will open an issue and tag you for thinking through that if you are game.

Good catch on the docs `item` vs `enum`. Will fix that up.

For replacing defaults in the latex export, we are using `myst-to-tex` and currently don't have a great way to expose that in yaml. I will open an issue with some pointers and thoughts that I would love your take on.

I am really glad that you appreciate the `jtex` tool, I know that @fwkoch has put about a bunch of effort into low level validation and error reporting across the stack, and he will appreciate that it is worth it. :)

For the no success message, I kinda agree there should be a message. Will issue/tag/pr that after this.

I am away next week, and with @fwkoch pulling out the myst-cli next week, it might be better to meet/work on this the week of the 17th?

**Reply by kai-tub** [reply]
> On a positive note, I am SUPER stoked by all the feedback, especially that knowing the constraints we put in about the template (flat directory) are a hard no from you. We can change that, and mostly did it for simplicity in the copying to temp files, etc.

Great! I would strongly prefer to structure my code with directories, especially when others should also be able to write templates and might not have the ability to change the beamer template to take the files from a different location.

>  but you probably may want more control. I will open an issue and tag you for thinking through that if you are game.

Sure! But I also want to set the correct expectations. I wrote the beamer template quite some time ago, and I never wrote any LaTeX packages, so I am not a TeX-Wizard. But I think I can spot some 'easy' requirements/corner cases that aren't that obvious. :smile: 

And on that note:

Another thing I realized is that I cannot 'disable' libraries to be imported.
I wrote the template to work with pdflatex _and_ lualatex_. As such, specific libraries will _only_ be imported if they are necessary for the engine. I could imagine that some libraries should not be _forced_ to be imported as they might not exist/work for the other engine.
I would say that 95% are definitely shared across engines, but there should be a way to configure when what libraries should be included.
Also, some packages (like `hyperref` if I remember correctly) _should_ not be imported by users if they are set-up within the theme.
If a theme has to configure something, there might be a strong reason to _not_ let the user re-import a package and mess up the configuration.
I know that I came across this issue during the development of the theme. If there is a need, I could dig through my code/commits to figure out what package was causing the problem.

> Good catch on the docs item vs enum. Will fix that up.

:+1:

> For replacing defaults in the latex export, we are using myst-to-tex and currently don't have a great way to expose that in yaml. I will open an issue with some pointers and thoughts that I would love your take on.

:+1: 

> I am away next week, and with @fwkoch pulling out the myst-cli next week, it might be better to meet/work on this the week of the 17th?

Sure, feel free to ping me on GitHub. Though my availability is rather 'spikey' for such kinds of projects.
If I am in the _mood_, I tend to _hyper-focus_ and communicate/write/test a lot (as I am currently doing across the executable-books project on accident :laughing: ) and then take a longer break (whatever that means).

**Reply by fwkoch** [reply]
👋 Hello! Thanks for all the great feedback! Really nice to have as I'm getting more into the `myst-cli` work.

**Reply by kai-tub** [reply]
@rowanc1 or @fwkoch if there is anything that you would like to discuss/get some feedback on or want me to try out, please let me know :)
I am happy to test things out if you ping me but I am not "subscribed" to your packages and will probably not notice any big updates otherwise. 😅

**Reply by rowanc1** [reply]
We are just doing a huge merge from curvenote to MyST today, and then want to resurrect this thread!

**Reply by kai-tub** [reply]
Off-topic:
I read this article and thought that it might be helpful for the future if mystjs will move _back_ to the Python word:
https://medium.com/@jonmmease/introducing-vlconvert-c763f0076e89

:)
